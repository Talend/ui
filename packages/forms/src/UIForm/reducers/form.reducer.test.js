import formReducer from './form.reducer';
import {
	TF_CREATE_FORM,
	TF_CHANGE_FORM,
	TF_REMOVE_FORM,
	TF_MUTATE_VALUE,
	TF_SET_ALL_ERRORS,
	TF_SET_PARTIAL_ERROR,
} from '../actions';

const formName = 'formName';
const jsonSchema = { jsonSchema: 'json' };
const uiSchema = [{ uiSchema: 'json' }];
const properties = { props: 'json' };
const errors = { field: 'errors' };

const oldState = {
	existingFormName: {
		jsonSchema: { jsonSchema: 'oldJson' },
		uiSchema: [{ uiSchema: 'oldJson' }],
		properties: { props: 'oldJson' },
		errors: { field: 'oldError' },
	},
};

describe('Form reducers', () => {
	describe('#TF_CREATE_FORM', () => {
		it('should not create existing form', () => {
			// given
			const action = {
				type: TF_CREATE_FORM,
				formName: 'existingFormName',
				jsonSchema,
				uiSchema,
				properties,
				errors,
			};

			// when
			const state = formReducer(oldState, action);

			// then
			expect(state).toBe(oldState);
		});

		it('should not create new form', () => {
			// given
			const action = {
				type: TF_CREATE_FORM,
				formName,
				jsonSchema,
				uiSchema,
				properties,
				errors,
			};

			// when
			const state = formReducer(oldState, action);

			// then
			expect(state).toMatchSnapshot();
		});
	});

	describe('#TF_CHANGE_FORM', () => {
		it('should not replace anything when form does not exist', () => {
			// given
			const action = {
				type: TF_CHANGE_FORM,
				formName,
			};

			// when
			const state = formReducer(oldState, action);

			// then
			expect(state).toBe(oldState);
		});

		it('should not replace anything when there is no provided replacement', () => {
			// given
			const action = {
				type: TF_CHANGE_FORM,
				formName: 'existingFormName',
			};

			// when
			const state = formReducer(oldState, action);

			// then
			expect(state).toBe(oldState);
		});

		it('should replace the forms configurations', () => {
			// given
			const action = {
				type: TF_CHANGE_FORM,
				formName: 'existingFormName',
				jsonSchema,
				uiSchema,
				properties,
				errors,
			};

			// when
			const state = formReducer(oldState, action);

			// then
			expect(state).toMatchSnapshot();
		});
	});

	describe('#TF_REMOVE_FORM', () => {
		it('should remove form', () => {
			// given
			const action = {
				type: TF_REMOVE_FORM,
				formName: 'existingFormName',
			};

			// when
			const state = formReducer(oldState, action);

			// then
			expect(state).toMatchSnapshot();
		});
	});

	describe('#TF_MUTATE_VALUE', () => {
		it('should mutate value', () => {
			// given
			const action = {
				type: TF_MUTATE_VALUE,
				formName: 'existingFormName',
				schema: {
					key: ['props'],
				},
				value: 'newJson',
			};

			// when
			const state = formReducer(oldState, action);

			// then
			expect(state).toMatchSnapshot();
		});
	});

	describe('#TF_SET_ALL_ERRORS', () => {
		it('should replace all errors', () => {
			// given
			const action = {
				type: TF_SET_ALL_ERRORS,
				formName: 'existingFormName',
				errors: {
					field1: 'new error 1',
					field2: 'new error 2',
				},
			};

			// when
			const state = formReducer(oldState, action);

			// then
			expect(state).toMatchSnapshot();
		});
	});

	describe('#TF_SET_PARTIAL_ERROR', () => {
		it('should set errors', () => {
			// given
			const action = {
				type: TF_SET_PARTIAL_ERROR,
				formName: 'existingFormName',
				errors: {
					field1: 'new error 1',
					field2: 'new error 2',
				},
			};

			// when
			const state = formReducer(oldState, action);

			// then
			expect(state).toMatchSnapshot();
		});
	});
});
