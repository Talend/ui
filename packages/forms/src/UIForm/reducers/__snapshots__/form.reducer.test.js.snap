// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form reducers #TF_CHANGE_FORM should replace the forms configurations 1`] = `
Object {
  "existingFormName": Object {
    "errors": Object {
      "field": "oldError",
    },
    "jsonSchema": Object {
      "jsonSchema": "oldJson",
    },
    "properties": Object {
      "props": "oldJson",
    },
    "uiSchema": Array [
      Object {
        "uiSchema": "oldJson",
      },
    ],
  },
}
`;

exports[`Form reducers #TF_CREATE_FORM should not create new form 1`] = `
Object {
  "existingFormName": Object {
    "errors": Object {
      "field": "oldError",
    },
    "jsonSchema": Object {
      "jsonSchema": "oldJson",
    },
    "properties": Object {
      "props": "oldJson",
    },
    "uiSchema": Array [
      Object {
        "uiSchema": "oldJson",
      },
    ],
  },
  "formName": Object {
    "errors": Object {
      "field": "errors",
    },
    "jsonSchema": Object {
      "jsonSchema": "json",
    },
    "properties": Object {
      "props": "json",
    },
    "uiSchema": Array [
      Object {
        "uiSchema": "json",
      },
    ],
  },
}
`;

exports[`Form reducers #TF_MUTATE_VALUE should mutate value 1`] = `
Object {
  "existingFormName": Object {
    "errors": Object {
      "field": "oldError",
    },
    "jsonSchema": Object {
      "jsonSchema": "oldJson",
    },
    "properties": Object {
      "props": "oldJson",
    },
    "uiSchema": Array [
      Object {
        "uiSchema": "oldJson",
      },
    ],
  },
}
`;

exports[`Form reducers #TF_REMOVE_FORM should remove form 1`] = `Object {}`;

exports[`Form reducers #TF_SET_ALL_ERRORS should replace all errors 1`] = `
Object {
  "existingFormName": Object {
    "errors": Object {
      "field1": "new error 1",
      "field2": "new error 2",
    },
    "jsonSchema": Object {
      "jsonSchema": "oldJson",
    },
    "properties": Object {
      "props": "oldJson",
    },
    "uiSchema": Array [
      Object {
        "uiSchema": "oldJson",
      },
    ],
  },
}
`;

exports[`Form reducers #TF_SET_PARTIAL_ERROR should set errors 1`] = `
Object {
  "existingFormName": Object {
    "errors": Object {
      "field": "oldError",
      "field1": "new error 1",
      "field2": "new error 2",
    },
    "jsonSchema": Object {
      "jsonSchema": "oldJson",
    },
    "properties": Object {
      "props": "oldJson",
    },
    "uiSchema": Array [
      Object {
        "uiSchema": "oldJson",
      },
    ],
  },
}
`;
